bot.ml

```diff
@@ -786,15 +786,119 @@ let callback _conn req body =
                   ()
           else if
             string_match ~regexp:(f "@%s: [Mm]erge now" bot_name) body
-            && Option.is_some comment_info.pull_request
-          then
-            (* Should be restricted to coq *)
+            && Bool.equal comment_info.issue.pull_request true
+            && String.equal comment_info.issue.issue.owner "coq"
+            && String.equal comment_info.issue.issue.repo "coq"
+            (*&&
+              let pr =
+                (Option.value_exn ~message:"Attempting to merge an issue."
+                   comment_info.pull_request)
+                  .issue
+              in
+              String.equal pr.issue.owner "coq"
+              && String.equal pr.issue.repo "coq"*)
+          then (
+            (fun () ->
+              (*let pr =
+                  Option.value_exn ~message:"Attempting to merge an issue."
+                    comment_info.pull_request
+                in
+                GitHub_queries.get_merge_pull_request_refs ~bot_info
+                  ~owner:pr.issue.issue.owner ~repo:pr.issue.issue.repo
+                  ~number:pr.issue.issue.number*)
+              let pr = comment_info.issue in
+              GitHub_queries.get_merge_pull_request_refs ~bot_info
+                ~owner:pr.issue.owner ~repo:pr.issue.repo
+                ~number:pr.issue.number
+              >>= function
+              | Ok merge_info -> (
+                  if
+                    List.exists merge_info.labels ~f:(fun label ->
+                        string_match ~regexp:"needs:.*" label)
+                  then
+                    GitHub_mutations.post_comment ~bot_info
+                      ~message:
+                        (f "@%s: Needs label found, merge aborted."
+                           comment_info.author)
+                      ~id:pr.id
+                  else if
+                    not
+                      (List.exists merge_info.labels ~f:(fun label ->
+                           string_match ~regexp:"kind:.*" label))
+                  then
+                    GitHub_mutations.post_comment ~bot_info
+                      ~message:
+                        (f "@%s: No kind label found, merge aborted."
+                           comment_info.author)
+                      ~id:pr.id
+                    (* merge_info already contains a valid milestone, no need to check if there's one *)
+                    (*else if
+                        List.exists merge_info.reviews ~f:(fun (_, state) ->
```

Comment:
To know the review status, instead of going through the entire list of reviews (which can even turn out to be several page long), you can just ask GitHub for the `reviewDecision` and fail if it is `REVIEW_REQUIRED` (missing an approve review from a maintainer) or `CHANGES_REQUESTED` (at least one non-outdated / non-dismissed change request).